src/java.base/share/classes/java/lang/Long.java

```diff
@@ -549,27 +551,39 @@ static int getChars(long i, int index, byte[] buf) {
         // Get 2 digits/iteration using longs until quotient fits into an int
         while (i <= Integer.MIN_VALUE) {
             q = i / 100;
-            r = (int)((q * 100) - i);
+            charPos -= 2;
+            UNSAFE.putShortUnaligned(
+                    buf,
+                    Unsafe.ARRAY_BYTE_BASE_OFFSET + charPos,
+                    Integer.PACKED_DIGITS[(int)((q * 100) - i)],
+                    false);
             i = q;
-            buf[--charPos] = Integer.DigitOnes[r];
-            buf[--charPos] = Integer.DigitTens[r];
         }
 
         // Get 2 digits/iteration using ints
         int q2;
         int i2 = (int)i;
         while (i2 <= -100) {
             q2 = i2 / 100;
-            r  = (q2 * 100) - i2;
+            charPos -= 2;
+            UNSAFE.putShortUnaligned(
+                    buf,
+                    Unsafe.ARRAY_BYTE_BASE_OFFSET + charPos,
+                    Integer.PACKED_DIGITS[(q2 * 100) - i2],
+                    false);
             i2 = q2;
-            buf[--charPos] = Integer.DigitOnes[r];
-            buf[--charPos] = Integer.DigitTens[r];
         }
 
         // We know there are at most two digits left at this point.
-        buf[--charPos] = Integer.DigitOnes[-i2];
         if (i2 < -9) {
-            buf[--charPos] = Integer.DigitTens[-i2];
+            charPos -= 2;
+            UNSAFE.putShortUnaligned(
+                    buf,
+                    Unsafe.ARRAY_BYTE_BASE_OFFSET + charPos,
+                    Integer.PACKED_DIGITS[-i2],
+                    false);
```

Comment:
Replace the implicit array bounds check with an explicit array index check if using Unsafe.

Reply:
I added assert with reference to the implementation of StringUTF16#putChar, is this safe enough?